### PR TITLE
Improve the `lovd_fixHGVS()` function.

### PR DESCRIPTION
**Pull request by loeswerkman**
Resolving the comments on the pull request of improve/getVariantInfo

### CODE REVIEW

**Review by ifokkema — CHANGES_REQUESTED** [review]
I fixed some issues myself (see new commits) and have marked other issues that need to be resolved.

File: tests/unit_tests/FixHGVS.php [hunk 1]
```diff
@@ -149,14 +149,34 @@ public static function dataProviderFixHGVS ()
 
             // FIXABLE VARIANTS.
             // Missing prefixes.
+            // > VARIANTS WHICH DON'T NEED FIXING.
```
Comment (ifokkema) [inline tests/unit_tests/FixHGVS.php]: These comments are in conflict.

File: tests/unit_tests/FixHGVS.php [hunk 2]
```diff
@@ -186,103 +206,56 @@ public static function dataProviderFixHGVS ()
...
             array('g.5_(10_?)del', 'g.5_(10_?)del'),
-            array('g.5_(?_10)del', 'g.5_(?_10)del'), // Fixme; have another look.
+            array('g.5_(?_10)del', 'g.5_(?_10)del'),
+            array('g.?_(?_10)del', 'g.(?_10)del'),
             array('g.5_(?_?)del', 'g.5_?del'),
+
             array('g.(?_5)_(10_?)del', 'g.(?_5)_(10_?)del'),
-            array('g.(5_?)_(?_10)del', 'g.(5_10)del'),
-            array('g.(?_?)_(?_?)del', 'g.?del'),
+            array('g.(5_?)_(?_10)del', 'g.(5_?)_(?_10)del'),
```
Comment (ifokkema) [inline tests/unit_tests/FixHGVS.php:235]: Why this change? The old example was correct, no?

File: tests/unit_tests/FixHGVS.php [hunk 3]
```diff
@@ -186,103 +206,56 @@ public static function dataProviderFixHGVS ()
...
+            array('g.(5_?)_(?_10)del(3)', 'g.(5_10)del(3)'),
```
Comment (ifokkema) [inline tests/unit_tests/FixHGVS.php:236]: This seems to be in conflict with the example two lines above this one. Is `g.(5_?)_(?_10)` replaced into `g.(5_10)` or not? Does it depend on the suffix? Although I can see the logic behind that, I also remember Johan saying `g.(5_?)_(?_10)` was always incorrect and it should be replaced to `g.(5_10)`. If `g.(5_10)del` is not correct HGVS, we should check with him just to be sure.

File: tests/unit_tests/GetVariantInfoTest.php [hunk 4]
```diff
@@ -332,7 +332,7 @@ public static function dataProviderGetVariantInfo ()
                 'position_end' => 2,
                 'type' => 'repeat',
                 'warnings' => array(
-                    'WNOTSUPPORTED' => 'Repeat variants are currently not supported for mapping and validation.',
+                    'WNOTSUPPORTED' => 'Repeat variants are currently not supported for mapping and validation, because external tools do not recognise them yet.',
```
Comment (ifokkema) [inline tests/unit_tests/GetVariantInfoTest.php:335]: Please remove this addition, as we can't align this with VV.

File: tests/unit_tests/GetVariantInfoTest.php [hunk 5]
```diff
@@ -343,7 +343,7 @@ public static function dataProviderGetVariantInfo ()
                 'position_end_intron' => 0,
                 'type' => 'repeat',
                 'warnings' => array(
-                    'WNOTSUPPORTED' => 'Repeat variants are currently not supported for mapping and validation.',
+                    'WNOTSUPPORTED' => 'Repeat variants are currently not supported for mapping and validation, because external tools do not recognise them yet.',
```
Comment (ifokkema) [inline tests/unit_tests/GetVariantInfoTest.php:346]: Please remove this addition, as we can't align this with VV.

File: tests/unit_tests/GetVariantInfoTest.php [hunk 6]
```diff
@@ -364,7 +364,7 @@ public static function dataProviderGetVariantInfo ()
                 'position_end' => 2,
                 'type' => 'repeat',
                 'warnings' => array(
-                    'WNOTSUPPORTED' => 'Repeat variants are currently not supported for mapping and validation.',
+                    'WNOTSUPPORTED' => 'Repeat variants are currently not supported for mapping and validation, because external tools do not recognise them yet.',
```
Comment (ifokkema) [inline tests/unit_tests/GetVariantInfoTest.php:367]: Please remove this addition, as we can't align this with VV.
(idem the other places where this has been added)

File: src/inc-lib-variants.php [hunk 7]
```diff
@@ -194,7 +195,7 @@ function lovd_fixHGVS ($sVariant, $sType = 'g')
                 return lovd_fixHGVS(
                     $sBeforeSuffix . $aVariantInfo['type'] . str_replace(array('(', ')'), '', $sSuffix), $sType);
 
-            } elseif (preg_match('/^\[[^NX][^;]+\]$/', $sSuffix)) {
+            } elseif (preg_match('/^\[[^NX][^;]*]$/', $sSuffix)) {
```
Comment (ifokkema) [inline src/inc-lib-variants.php:198]: Why was this changed back? Did I not understand what kind of possible values are matched here?

By the way, `]` does not strictly need an `\` in front of it, but I find it easily confused with the `]` which means the end of a range of characters that can be expected. So I prefer escaping `]` the same way `[` needs to be escaped.

File: src/inc-lib-variants.php [hunk 8]
```diff
@@ -269,12 +270,58 @@ function lovd_fixHGVS ($sVariant, $sType = 'g')
...
+                $aPositions['C'] = '';
+                $sBefore = $sBefore . '(';
+
+            } elseif ($aPositions['B'] . $aPositions['C'] == '??' && !$aPositions['D']) {
+                // e.g. c.(1_?)_?del
+                $aPositions['B'] = '';
+                $sAfter = ')' . $sAfter;
+
+            } elseif ($aPositions['B'] . $aPositions['C'] == '??' && $aPositions['A'] != '?' &&
+                      !in_array($aPositions['D'], array('', '?'))) {
```
Comment (ifokkema) [inline src/inc-lib-variants.php:304]: Whenever an if needs to be split in multiple lines, I find it more readable to put the `&&` or `||` on the new line instead of the end of the first line. On first glance, having the `&&` or `||` at the beginning of the new line makes it directly clear that line still belongs to the `if()` condition. Having it at the end of the previous line requires you to first look at whether or not the second line is the body of the `if()` or still the condition itself.

File: tests/unit_tests/GetVariantInfoTest.php [hunk 9]
```diff
@@ -648,7 +651,7 @@ public static function dataProviderGetVariantInfo ()
                 'position_end' => 2,
                 'type' => 'ins',
                 'warnings' => array(
-                    'WSUFFIXFORMAT' => 'The inserted sequence does not follow HGVS guidelines.',
+                    'WSUFFIXFORMAT' => 'The inserted/affected sequence does not follow HGVS guidelines.',
```
Comment (ifokkema) [inline tests/unit_tests/GetVariantInfoTest.php:654]: Does it say "inserted/affected" because, in the code, you're not sure whether you're handling insertions or something else at that time? Or because you think "inserted sequence" isn't clear enough for an insertion?

File: tests/unit_tests/FixHGVS.php [hunk 10]
```diff
@@ -41,30 +41,123 @@ public function testFixHGVS ($sInput, $sOutput)
...
+            array('c.1_2AC[20]', 'c.1_2AC[20]'),
+            array('c.1_2A>G', 'c.1_2A>G'),
+            array('g.=', 'g.='),
+            array('c.1insA', 'c.1insA'),
+            array('c.1_2ins', 'c.1_2ins'),
+            array('c.1_10insA', 'c.1_10insA'),
+            array('c.(1_2)insA', 'c.(1_2)insA'),
+            array('c.1_20insBLA', 'c.1_20insBLA'),
+            array('c.1_100insA', 'c.1_100insA'),
+            array('c.1_100del(10)', 'c.1_100del'), // Fixme; take another look!!
```
Comment (ifokkema) [inline tests/unit_tests/FixHGVS.php:159]: This variant is listed as unfixable but is actually edited and fixed. Is it supposed to be there?

File: src/inc-lib-variants.php [hunk 11]
```diff
@@ -45,100 +45,279 @@ function lovd_fixHGVS ($sVariant, $sType = 'g')
...
+    
+    } elseif (!empty($aVariantInfo['errors'])) {
+        return $sVariant; // Not HGVS.
+    }
+    
+    
+    // Change the variant type (if possible) if the wrong type was chosen.
+    if (isset($aVariantInfo['warnings']['WWRONGTYPE'])) {
+        if ($aVariantInfo['type'] == 'subst') {
+            return lovd_fixHGVS(preg_replace('/[ACTG]>/', 'delins', $sVariant), $sType);
```
[the remark this answers is not in the document]
Comment (ifokkema) [reply src/inc-lib-variants.php]: Ik heb deze bug nu opgelost.

File: src/inc-lib-variants.php [hunk 12]
```diff
@@ -45,100 +45,279 @@ function lovd_fixHGVS ($sVariant, $sType = 'g')
...
-        // Also convert U to T, since lowercase bases may mean an RNA-based description.
-        return lovd_fixHGVS($aRegs[1] . strtoupper(str_replace('u', 't', $aRegs[2])));
+    
+    // Remove the suffix if it is given to a variant type which should not hold one.
+    if (isset($aVariantInfo['warnings']['WSUFFIXGIVEN'])) {
+        // The warning message stores the spot of the variant after which the suffix is given.
+        // We find this spot by taking the part within the double quotes.
+        preg_match('/\".+\"/', $aVariantInfo['warnings']['WSUFFIXGIVEN'], $aMatches);
+        $sBeforeSuffix = str_replace('"', '', $aMatches[0]);
+        return lovd_fixHGVS(explode($sBeforeSuffix, $sVariant)[0] . $sBeforeSuffix, $sType);
```
[the remark this answers is not in the document]
Comment (ifokkema) [reply src/inc-lib-variants.php]: Ik heb dit nu gefixed.

File: src/inc-lib-variants.php [hunk 13]
```diff
@@ -45,100 +45,279 @@ function lovd_fixHGVS ($sVariant, $sType = 'g')
...
+    
+    // Replace uracil with thymine (RNA -> DNA description).
+    if (strpos($sVariant, 'U') !== false) {
+        return lovd_fixHGVS(str_replace('dTp', 'dup', str_replace('U', 'T', $sVariant)), $sType);
+    }
+    
+    // Make sure no unnecessary bases are given for wild types (c.123A= -> c.123=).
+    if (strpos($sVariant, '=') !== false) {
+        return lovd_fixHGVS(str_replace(array('=', 'A', 'C', 'T', 'G'), '', $sVariant) . '=', $sType);
+    }
```
[the remark this answers is not in the document]
Comment (ifokkema) [reply src/inc-lib-variants.php]: Verified this bug; also, when submitting `c.(123A=)`, `lovd_getVariantInfo()` returns a lot of notices.

File: src/inc-lib-variants.php [hunk 14]
```diff
@@ -45,100 +45,279 @@ function lovd_fixHGVS ($sVariant, $sType = 'g')
...
     }
-
-    return $sVariant;
+    
+    
+    // Reformat wrongly described suffixes.
+    if (isset($aVariantInfo['warnings']['WSUFFIXFORMAT'])) {
+        list($sBeforeSuffix, $sSuffix) = explode($aVariantInfo['type'], $sVariant);
+
+        if (preg_match('/^[0-9]*$/', $sSuffix)) {
```
[the remark this answers is not in the document]
Comment (ifokkema) [reply src/inc-lib-variants.php]: Replaced this with a `ctype_digit()`.

File: src/inc-lib-variants.php [hunk 15]
```diff
@@ -45,100 +45,279 @@ function lovd_fixHGVS ($sVariant, $sType = 'g')
...
+        
+        if (($aPositions['C'] &&
+             $aPositions['A'] + ($aPositions['B']? : $aPositions['A']) >
+             $aPositions['C'] + ($aPositions['D']? : $aPositions['C']))) {
+            // If this is the case, the positions are swapped in groups,
+            //  i.e., c.(6_10)_(1_5)del. We will fix this as follows:
+            list($aPositions['A'], $aPositions['B'], $aPositions['AIntron'], $aPositions['BIntron'],
+                 $aPositions['C'], $aPositions['D'], $aPositions['CIntron'], $aPositions['DIntron']) =
+                array($aPositions['C'], $aPositions['D'], $aPositions['CIntron'], $aPositions['DIntron'],
+                      $aPositions['A'], $aPositions['B'], $aPositions['AIntron'], $aPositions['BIntron']);
```
[the remark this answers is not in the document]
Comment (ifokkema) [reply src/inc-lib-variants.php]: This code also seems to contain some bugs; some tests are failing and I believe that's due to this code. Perhaps we should discuss this in person (Thu or Fri).

File: src/inc-lib-variants.php [hunk 16]
```diff
@@ -45,100 +45,279 @@ function lovd_fixHGVS ($sVariant, $sType = 'g')
...
+    //  and fix the variant.
+    $aVariantInfo = lovd_getVariantInfo($sVariant, false);
+    if ($aVariantInfo === false) {
+        return $sVariant; // Not HGVS.
+    
+    } elseif (isset($aVariantInfo['errors']['EFALSEUTR']) || isset($aVariantInfo['errors']['EFALSEINTRONIC'])) {
+        // The wrong prefix was given. In other words: intronic positions or UTR
+        //  notations were found for genomic DNA.
+        // Fixme; take another look.
+        $sType = 'c';
```
[the remark this answers is not in the document]
Comment (ifokkema) [reply src/inc-lib-variants.php:154]: Perhaps we should rely on the given `$sType`?

File: src/inc-lib-variants.php [hunk 17]
```diff
@@ -1,540 +1,514 @@
...
-}
-?>
+<?php /** @noinspection PhpStrFunctionsInspection */
+/*******************************************************************************
+ *
+ * LEIDEN OPEN VARIATION DATABASE (LOVD)
+ *
+ * Created     : 2016-01-22
+ * Modified    : 2021-11-10
+ * For LOVD    : 3.0-28
```
[the remark this answers is not in the document]
Comment (ifokkema) [reply src/inc-lib-variants.php]: Fixed this myself, but keeping this comment open as a reminder.

File: src/inc-lib-init.php [hunk 18]
```diff
@@ -1334,6 +1334,7 @@ function lovd_getVariantInfo ($sVariant, $sTranscriptID = '', $bCheckHGVS = fals
                  array('earliest_start',   'latest_start'),
                  array('earliest_end',     'latest_end'),
                  array('earliest_start',   'earliest_end'),
+                 array('latest_start',     'earliest_end'),
```
Comment (loeswerkman) [inline src/inc-lib-init.php:1337]: This additional check makes sure that variants such as c.(1_3)_3del are now also being looked at.

File: src/inc-lib-init.php [hunk 19]
```diff
@@ -1345,16 +1346,17 @@ function lovd_getVariantInfo ($sVariant, $sTranscriptID = '', $bCheckHGVS = fals
             $sIntronicLast  = str_replace('_', '_intronic_', $sLast);
 
             if ($aVariant[$sFirst] > $aVariant[$sLast]) {
-                list($aVariant[$sFirst], $aVariant[$sLast]) = array($aVariant[$sLast], $aVariant[$sFirst]);
+                list($aVariant[$sFirst], $aVariant[$sIntronicFirst], $aVariant[$sLast], $aVariant[$sIntronicLast]) =
+                    array($aVariant[$sLast], $aVariant[$sIntronicLast], $aVariant[$sFirst], $aVariant[$sIntronicFirst]);
```
Comment (loeswerkman) [inline src/inc-lib-init.php:1350]: This fixes the bug which failed to switch the intronic values along with the 'basic' values.

File: tests/unit_tests/GetVariantInfoTest.php [hunk 20]
```diff
@@ -143,7 +143,7 @@ public static function dataProviderGetVariantInfo ()
                 'type' => 'subst',
                 'warnings' => array(),
                 'errors' => array(
-                    'EPOSITIONFORMAT' => 'Too many positions are given for variant type substitution.'
+                    'ETOOMANYPOSITIONS' => 'Too many positions are given for variant type substitution.'
```
Comment (loeswerkman) [inline tests/unit_tests/GetVariantInfoTest.php:146]: By giving these cases their own unique error ID, fixHGVS can easily spot and repair them.

File: src/inc-lib-init.php [hunk 21]
```diff
@@ -1560,8 +1562,13 @@ function lovd_getVariantInfo ($sVariant, $sTranscriptID = '', $bCheckHGVS = fals
...
             }
-            $aResponse['errors']['EPOSITIONFORMAT'] = 'Too many positions are given for variant type substitution.';
-            return $aResponse;
+            if ($aVariant['earliest_start'] != $aVariant['earliest_end']) {
+                // If the two positions are not the same, the variant is not fixable.
+                $aResponse['errors']['ETOOMANYPOSITIONS'] = 'Too many positions are given for variant type substitution.';
+                return $aResponse;
+            }
+            // If the positions are the same, the variant can safely be interpreted and fixed accordingly.
+            $aResponse['warnings']['WTOOMANYPOSITIONS'] = 'Too many positions are given for variant type substitution.';
```
Comment (loeswerkman) [inline src/inc-lib-init.php:1571]: By returning a warning if both positions are the same, and an error if the two positions are different, fixHGVS can easily recognise what she can and cannot do to fix this variant.

Comment (ifokkema) [reply src/inc-lib-init.php:1571]: This entire block (from "Checking all type-specific format requirements" to the last `elseif()`) really needs comments (in the code). I'm getting lost there. There's so many `if()s` on the position fields, that I can't see anymore what kind of variants each block is aiming for. I can deduce it from the error messages, but having examples documented in the code makes it easier to spot the relevant blocks, plus, it becomes easier to spot possible omissions.

File: src/inc-lib-variants.php [hunk 22]
```diff
@@ -150,9 +150,27 @@ function lovd_fixHGVS ($sVariant, $sType = 'g')
...
-        return lovd_fixHGVS($sType . substr($sVariant, 1), $sType);
+        if ($sVariant[0] == $sType) {
+            if (isset($aVariantInfo['errors']['EFALSEUTR'])
+                || ($aVariantInfo['position_start'] < 250000 && $aVariantInfo['position_start_intronic'] < 250000)) {
+                // If the prefix equals the expected type, there is nothing
+                //  much that we can do about receiving a false UTR.
+                // If variants hold false intronic positions, it might be that
+                //  the user accidentally wrote down '-' while meaning '_'.
+                // We will fix this only if we can be really sure this is the case,
+                //  which is if the position indicates a length too big to
```
Comment (ifokkema) [inline src/inc-lib-variants.php:161]: "the position indicates a length" should be "the variant contains a position".

**Review by ifokkema — CHANGES_REQUESTED** [review]
Requiring some more documentation in the code, and found one `if()` that doesn't look right to me.

File: tests/unit_tests/FixHGVS.php [hunk 23]
```diff
@@ -49,22 +49,239 @@ public static function dataProviderFixHGVS ()
     {
         // Data provider for testFixHGVS().
         return array(
-            // HGVS OK, returns input.
-            array('g.123A>C', 'g.123A>C'),
-            array('g.123del', 'g.123del'),
```
[the remark this answers is not in the document]
Comment (loeswerkman) [reply tests/unit_tests/FixHGVS.php:54]: Yes, they can all be found somewhere along the test, following my new structure for this test case

File: tests/unit_tests/FixHGVS.php (hunk 10 above)
Comment (loeswerkman) [reply tests/unit_tests/FixHGVS.php:163]: It is added to the test to make note of the fact that this happens, but I meant to talk through this one. We can do so later today :)

File: tests/unit_tests/GetVariantInfoTest.php (hunk 9 above)
Comment (loeswerkman) [reply tests/unit_tests/GetVariantInfoTest.php:654]: It says instered/affected because this message is now used not only for insertions, but also for suffixes for other type of variants, such as in the case of c.1_2del[A]

File: src/inc-lib-variants.php [hunk 24]
```diff
@@ -150,9 +150,27 @@ function lovd_fixHGVS ($sVariant, $sType = 'g')
     } elseif (isset($aVariantInfo['errors']['EFALSEUTR']) || isset($aVariantInfo['errors']['EFALSEINTRONIC'])) {
         // The wrong prefix was given. In other words: intronic positions or UTR
         //  notations were found for genomic DNA.
-        // FIXME; This is not always the case; see variants like g.157097179-157100787del.
-        $sType = 'c';
-        return lovd_fixHGVS($sType . substr($sVariant, 1), $sType);
+        if ($sVariant[0] == $sType) {
+            if (isset($aVariantInfo['errors']['EFALSEUTR'])
+                || ($aVariantInfo['position_start'] < 250000 && $aVariantInfo['position_start_intronic'] < 250000)) {
```
[the remark this answers is not in the document]
Comment (loeswerkman) [reply src/inc-lib-variants.php]: Let's talk through this when we meet at 10.15 :)

File: src/inc-lib-init.php (hunk 21 above)
Comment (loeswerkman) [reply src/inc-lib-init.php:1637]: Good point! Has been done :)

File: src/inc-lib-init.php [hunk 25]
```diff
@@ -1113,28 +1113,53 @@ function lovd_getVariantInfo ($sVariant, $sTranscriptID = '', $bCheckHGVS = fals
...
+
+            if ($sNCBIID == strstr($sVariant, ':', true)) {
+                // The transcript given in the DNA description is also
+                //  the transcript that we're using in LOVD for this variant.
+                $aResponse['warnings']['WTRANSCRIPTFOUND'] = 'Transcript ID found in the DNA description.';
+            } else {
+                // This is an actual problem; the submitter used a
+                //  different transcript than configured in LOVD.
+                $aResponse['warnings']['WDIFFERENTTRANSCRIPT'] =
+                    'The transcript found in the DNA description does not match the configured transcript.';
```
Comment (ifokkema) [inline src/inc-lib-init.php]: This is actually not true for NC(NM) notation. This code may need some work, but we can do that later.

File: src/inc-lib-init.php [hunk 26]
```diff
@@ -1113,28 +1113,53 @@ function lovd_getVariantInfo ($sVariant, $sTranscriptID = '', $bCheckHGVS = fals
...
+            }
         }
 
-        $sVariant = substr(strstr($sVariant, ':'), 1);
+        list($sReferenceSequence, $sVariant) = explode(':', $sVariant);
```
Comment (ifokkema) [inline src/inc-lib-init.php]: Please take note that `explode()` will just keep splitting the string. If `$sVariant` contains more colons, you'll lose that part of the variant. I'll fix this code by limiting `explode()` to a max of two parts.

**Review by ifokkema — CHANGES_REQUESTED** [review]
Some changes are needed, but I'll take care of them.

File: src/inc-lib-variants.php [hunk 27]
```diff
@@ -62,11 +70,13 @@ function lovd_fixHGVS ($sVariant, $sType = 'g', $sReference = '')
     // Check for a reference sequence.
     if ($sReference !== false) {
         if ($sReference == '') {
-            if (preg_match('/^[NX][CGMR]_[0-9]{6,9}\.[0-9]+:/', $sVariant)) {
+            if (preg_match('/^(\(?(LRG|ENS[GT]|[NX][CGMRTW])_?[0-9]+([.t][0-9]+)?\)?){1,2}:/', $sVariant)) {
```
Comment (ifokkema) [inline src/inc-lib-variants.php:73]: This regexp allows for invalid reference sequences to be passed. As the function doesn't do anything with the reference (`lovd_getVariantInfo()` checks it, I guess), maybe we don't need to be too strict about it. I'll have a look.

File: src/inc-lib-variants.php [hunk 28]
```diff
@@ -62,11 +70,13 @@ function lovd_fixHGVS ($sVariant, $sType = 'g', $sReference = '')
...
                 list($sReference, $sVariant) = explode(':', $sVariant);
-                $sReference .= ':'; // To easy the concatenation.
+                $sReference .= ':'; // We add the ':' to ease the concatenation later on.
             } else {
-                // No reference was found.
+                // No reference was found. We fill this in by false so
+                //  the function will not have to continue to do this
+                //  test every time the function calls upon itself.
```
Comment (ifokkema) [inline src/inc-lib-variants.php:79]: Looking at the code, I didn't think the function could still call itself well, so I tried to add a test for it. But I keep getting weird results and even an infinite loop. I have to investigate, I'm not sure what's going on.

**Review by ifokkema — CHANGES_REQUESTED** [review]
Strange issues when testing variants that cause the function to call itself.

File: tests/unit_tests/FixHGVS.php (hunk 23 above)
[the remark this answers is not in the document]
Comment (ifokkema) [reply tests/unit_tests/FixHGVS.php:54]: Actually, I'm afraid they're really deleted. I can't find them anywhere with a search. I'll re-add them!

File: tests/unit_tests/FixHGVS.php (hunk 10 above)
Comment (ifokkema) [reply tests/unit_tests/FixHGVS.php:163]: I see it's not updated (as in, `lovd_fixHGVS()` still removes the suffix). I can't find this in your Trello board, but we discussed not removing the suffix in this case, right?

File: tests/unit_tests/GetVariantInfoTest.php (hunk 9 above)
Comment (ifokkema) [reply tests/unit_tests/GetVariantInfoTest.php:654]: Ah, clear, thanks! I'll add an `if()` then! It'll clarify things for the user and doesn't make it much harder for us.

**Review by ifokkema — COMMENTED** [review]
`inc-lib-init.php` has been completely rewritten again, changing the LFs to CRLFs. Please check the diff before you commit your changes, and try to determine how you make this switch.

I will reverse this change.

File: src/inc-lib-init.php [hunk 29]
```diff
@@ -1157,8 +1157,8 @@ function lovd_getVariantInfo ($sVariant, $sTranscriptID = '', $bCheckHGVS = fals
 
         '([ACGT]+>[ACGTRYSWKMBDHVN]+|' .         //  | (substitution)
         '([ACTG]+\[[0-9]+])+' .                  //  | (repeat sequence)
-        '|ins|dup|delins|del|inv|[=?]' .         //  V
-        '|\|(gom|lom|met=))' .                   // 20. Type of variant
+        '|ins|dup|delins|del|inv|sup|[=?]' .     //  V
+        '|\|(gom|lom|met=|.{3}))' .              // 20. Type of variant
```
[the remark this answers is not in the document]
Comment (ifokkema) [reply src/inc-lib-init.php:1221]: Something is still odd here. Sending in `g.123|bsrC`; (there's no test for this variant type yet), we get:
```
'warnings' => Array (
    'WSUFFIXGIVEN' => 'Nothing should follow "|bsr".'
)
```

I'm tempted to delay that issue until later.

File: src/inc-lib-init.php [hunk 30]
```diff
@@ -1248,8 +1248,10 @@ function lovd_getVariantInfo ($sVariant, $sTranscriptID = '', $bCheckHGVS = fals
             if (strpos($sVariant, $sNeedsAPipe) && !strpos($sVariant, '|' . $sNeedsAPipe)) {
                 $aResponse['warnings']['WPIPEMISSING'] =
                     'Please place a "|" between the positions and the variant type (' . $sNeedsAPipe . ').';
+                return $aResponse;
             }
         }
+        return false;
```
[the remark this answers is not in the document]
Comment (ifokkema) [reply src/inc-lib-init.php:1276]: Actually, it just led to notices and thus failing tests ;)

File: src/inc-lib-variants.php [hunk 31]
```diff
@@ -224,7 +230,18 @@ function lovd_fixHGVS ($sVariant, $sType = '')
...
+            //  length of the variant. If we find this in variants of which the
+            //  positions are '[0-9]+_[0-9]+', it might be that the user forgot
+            //  to use brackets around the positions (indicating an uncertain
+            //  location, which would mean the suffix IS necessary). We cannot
+            //  be sure we may remove it, so we have to let this mistake rest.
+            $sSuffix = preg_replace('/.*' . $sVariantType . '/', '', $sVariant);
+            return lovd_fixHGVS(
+                $sReference . strstr($sVariant, $sVariantType, true) . $sVariantType .
+                (substr_count($sSuffix, '(') >= substr_count($sSuffix, ')')? '' : ')'), $sType);
+        }
```
[the remark this answers is not in the document]
Comment (ifokkema) [reply src/inc-lib-variants.php:237]: I simplified these lines a bit, but they still behave the same.

File: src/inc-lib-init.php (hunk 29 above)
[the remark this answers is not in the document]
Comment (ifokkema) [reply src/inc-lib-init.php]: OK, I fixed this in 7bbe366e.

**Comment by ifokkema** [comment]
@loeswerkman I'm not done yet, but I'll stop for today. Before you make your next commit, please stash your local changes, pull in these changes, and unstash your changes. If there are conflicts and you need help resolving them, let me know!

File: src/inc-lib-init.php [hunk 32]
```diff
@@ -1151,8 +1151,8 @@ function lovd_getVariantInfo ($sVariant, $sTranscriptID = '', $bCheckHGVS = fals
 
             $sReferenceType = preg_replace('/[0-9_.-]/', '', $sReferenceSequence);
 
-            if (($sVariant[0] == 'n' && !preg_match('/(NR|N[CGTW]\(NR\)|ENST|LRGt?)/', $sReferenceType))
-                || ($sVariant[0] == 'c' && !preg_match('/^([NX]M|N[CGTW]\(NM\)|ENST|LRGt?)$/', $sReferenceType))
+            if (($sVariant[0] == 'n' && !preg_match('/(NR|ENST|LRGt?)/', $sReferenceType))
+                || ($sVariant[0] == 'c' && !preg_match('/([NX]M|ENST|LRGt?)/', $sReferenceType))
```
Comment (ifokkema) [inline src/inc-lib-init.php:1155]: Can you explain to me this change? I'm tempted to reverse it?

File: src/inc-lib-init.php [hunk 33]
```diff
@@ -1175,6 +1175,16 @@ function lovd_getVariantInfo ($sVariant, $sTranscriptID = '', $bCheckHGVS = fals
                             ' For ' . $sVariant[0] . '. variants, please use a genomic reference sequence.';
                         break;
                 }
+
+            } elseif (!preg_match('/^(N[CGT]|LRG|ENSG)/', $sReferenceType)
```
Comment (ifokkema) [inline src/inc-lib-init.php:1179]: Somehow my earlier addition of NT and NW reference sequences was deleted here. It's now sometimes allowed and sometimes not, which seems inconsistent. It's good that LRG and ENSG got added, was the removal of NT and NW an accident or a deliberate change?

**Review by ifokkema — CHANGES_REQUESTED** [review]
We're almost there, I think!

File: src/inc-lib-init.php [hunk 34]
```diff
@@ -1744,12 +1750,13 @@ function lovd_getVariantInfo ($sVariant, $sTranscriptID = '', $bCheckHGVS = fals
...
 
-
     } elseif ($aVariant['suffix']) {
-        if (!in_array($aVariant['type'], array('ins', 'delins')) &&
-            (!isset($aResponse['messages']['IPOSITIONRANGE'])) || $aResponse['type'] == 'repeat') {
-            // If the variants are of a type that is not ins or delins, they should not have a suffix,
-            //  except if they were given a position range. Repeats should never be given a suffix.
+        if ($aResponse['type'] == 'repeat'
+            || (!in_array($aVariant['type'], array('ins', 'delins'))
+                && !isset($aResponse['messages']['IPOSITIONRANGE']))) {
```
Comment (ifokkema) [inline src/inc-lib-init.php:1756]: @loeswerkman, could you please check if I did this right? The `if()` condition was mixing `&&` and `||` without parentheses, and I think this is what you meant. I updated the order as well to aid in readability.

File: src/inc-lib-init.php [hunk 35]
```diff
@@ -1516,9 +1516,14 @@ function lovd_getVariantInfo ($sVariant, $sTranscriptID = '', $bCheckHGVS = fals
...
-        //  are uncertain, but somewhere within the range as given within the parentheses. We add a
-        //  message to make sure users know our interpretation and can make sure they meant it as such.
+        // We now know we are dealing with a case such as g.(1_5)ins. This means
+        //  that the positions are uncertain, but somewhere within the range as
+        //  given within the parentheses. We add a message to make sure users
+        //  know our interpretation and can make sure they meant it as such.
+        // Note that IPOSITIONRANGE and IUNCERTAINPOSITIONS both send the same
+        //  message to the user about uncertain positions. However, internally,
+        //  this notice is used to determine whether the variant needs a suffix
+        //  because the variant's position is a single, uncertain range.
```
Comment (ifokkema) [inline src/inc-lib-init.php:1526]: I decided in the end not to merge `IPOSITIONRANGE` and `IUNCERTAINPOSITIONS`, it wouldn't make the code more readable.

**Review by ifokkema — COMMENTED** [review]
@loeswerkman, could you please check this change? I'm not sure if I edited it right.
The only thing left now is one failing test. It already failed, so I'm not very sure how to fix it. If you know from the top of your head, please let me know. Otherwise, I'll have a look myself tomorrow.

**Review by ifokkema — APPROVED** [review]

**Comment by ifokkema** [comment]
Manually merged due to conflicts with `improve/getVariantInfo`.
